titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/KubeNotificationProcessor.java

```diff
@@ -316,7 +318,38 @@ private Task fillInMissingStates(V1Pod pod, Task task) {
         newStatusHistory.sort(Comparator.comparing(ExecutableStatus::getState));
 
         return task.toBuilder().withStatusHistory(newStatusHistory).build();
+    }
+
+    /**
+     * In case container could not be started, we do not have the container start time, only the finished time.
+     * The {@link V1PodStatus#getPhase()} is failed, and the {@link V1PodStatus#getMessage()} contains details on
+     * the nature of failure. There should be no launched state, so we add it to mark the container start attempt.
+     */
+    private Task fillInMissingStatesForContainerSetupFailure(V1Pod pod, Task task) {
+        // Sanity check. Should never be true.
+        if (JobFunctions.findTaskStatus(task, TaskState.Launched).isPresent()) {
+            return task;
+        }
+
+        long startAtTimestamp;
+        V1ContainerState containerState = KubeUtil.findContainerState(pod).orElse(null);
+        if (containerState != null && containerState.getTerminated() != null && containerState.getTerminated().getFinishedAt() != null) {
+            startAtTimestamp = containerState.getTerminated().getFinishedAt().getMillis();
+        } else {
+            startAtTimestamp = task.getStatus().getTimestamp();
+        }
+
+        List<TaskStatus> newStatusHistory = new ArrayList<>(task.getStatusHistory());
+        newStatusHistory.add(TaskStatus.newBuilder()
+                .withState(TaskState.Launched)
+                .withReasonCode(TaskStatus.REASON_STATE_MISSING)
+                .withReasonMessage("Filled in")
+                .withTimestamp(startAtTimestamp)
+                .build()
+        );
+        newStatusHistory.sort(Comparator.comparing(ExecutableStatus::getState));
```

Comment:
minor: can probably extract the comparator here as a constant to avoid allocations